### PR TITLE
[FEAT/TGUI] Better Gas Analyzer AND TGUI!

### PR DESCRIPTION
**Pull request by larentoun**
## What Does This PR Do

New:
- Adds Gas Analyzer TGUI
- Gas Analyzer provides tiny bit more info
- Gas Analyzer can be now used on turfs
- Gas Analyzer TGUI history with kPa/mol switch

Other:
- Gas Analyzer TGUI has toggleable auto-scan
- Gas scan messages no longer spam into the chat

## Why It's Good For The Game

Better visuals, less spam in chat

## Images of changes

https://github.com/ParadiseSS13/Paradise/assets/31931237/4e298f89-227d-4875-a74f-81dacfd16655

## Testing

See video

## Changelog
:cl:
add: Gas Analyzer TGUI
add: More info in Gas Analyzer
tweak: Ghost Gas Scan can now click on objects instead of scanning surroundings
del: Gas Scan messages in the chat when using Gas Analyzer
/:cl:

### CODE REVIEW

**Comment by Sirryan2002** [comment]
please post a working image of the TGUI

**Comment by Gatchapod** [comment]
How does this affect analyzing anomalies, if at all?

**Comment by larentoun** [comment]
> please post a working image of the TGUI

The video isn't enough?

**Comment by larentoun** [comment]
> How does this affect analyzing anomalies, if at all?

It should not affect them

**Comment by Sirryan2002** [comment]
The link seemed to be broken when the PR was opened

**Comment by TrashDoxx** [comment]
:trollface:

**Comment by 1080pCat** [comment]
does the history get saved through menu closing? if so does it start to lag on opening if you've scanned hundreds of tiles?

File: code/__HELPERS/atmospherics.dm [hunk 1]
```diff
@@ -0,0 +1,51 @@
+/** A simple rudimentary gasmix to information list converter. Can be used for UIs.
```
Comment (lewcc) [inline code/__HELPERS/atmospherics.dm:1]: This file needs to be renamed, perhaps to `atmospherics_helpers.dm`. It's failing linters

**Review by lewcc — CHANGES_REQUESTED** [review]
Please address merge conflicts and linters

**Review by S34NW — APPROVED** [review]
Deconflict for merge

File: code/game/objects/items/devices/scanners/gas_analyzer.dm [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+		if("modekpa")
+			history_view_mode = ANALYZER_HISTORY_MODE_KPA
+			return TRUE
+		if("modemol")
+			history_view_mode = ANALYZER_HISTORY_MODE_MOL
+			return TRUE
+
+/// Called when our analyzer is used on something
+/obj/item/analyzer/proc/on_analyze(datum/source, atom/target, save_data=TRUE)
+	SIGNAL_HANDLER
```
Comment (SteelSlayer) [inline code/game/objects/items/devices/scanners/gas_analyzer.dm:148]: This isn't called via a signal. Please clean up the arg from this and from all calls of the proc.
```suggestion
/obj/item/analyzer/proc/on_analyze(atom/target, save_data = TRUE)
```

File: code/modules/pda/utilities.dm [hunk 3]
```diff
@@ -117,28 +117,5 @@
...
-			pda.atmosanalyzer_scan(T.ptank.air_contents, user, T)
-	else if(istype(A, /obj/machinery/atmospherics/portable/scrubber/huge))
-		var/obj/machinery/atmospherics/portable/scrubber/huge/T = A
-		pda.atmosanalyzer_scan(T.air_contents, user, T)
-	else if(istype(A, /obj/machinery/atmospherics/unary/tank))
-		var/obj/machinery/atmospherics/unary/tank/T = A
-		pda.atmosanalyzer_scan(T.air_contents, user, T)
-
+/datum/data/pda/utility/scanmode/gas/scan_atom(atom/A, mob/user)
+	atmos_scan(user=user, target=A, silent=FALSE, print=TRUE)
```
Comment (SteelSlayer) [inline code/modules/pda/utilities.dm:121]:
```suggestion
	atmos_scan(user = user, target = A, silent = FALSE, print = TRUE)
```

File: code/__HELPERS/atmospherics_helpers.dm [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+ * -- Key: carbon_dioxide	Value: Number		  Desc: Amount of mols of CO2
+ * -- Key: nitrogen			Value: Number		  Desc: Amount of mols of N2
+ * -- Key: toxins			Value: Number		  Desc: Amount of mols of plasma
+ * -- Key: sleeping_agent	Value: Number		  Desc: Amount of mols of N2O
+ * -- Key: agent_b			Value: Number		  Desc: Amount of mols of agent B
+ * -- Key: total_moles		Value: Number		  Desc: Total amount of mols in the mixture
+ * Returned list should always be filled with keys even if value are nulls.
+ */
+
+//TODO: Port gas_mixture_parser from TG
```
Comment (SteelSlayer) [inline code/__HELPERS/atmospherics_helpers.dm:21]: Is this something you were going to do in this PR?

File: code/_onclick/observer_onclick.dm [hunk 5]
```diff
@@ -62,7 +62,10 @@
 /mob/dead/observer/ShiftClickOn(atom/A)
 	examinate(A)
 
-/atom/proc/attack_ghost(mob/user)
+/atom/proc/attack_ghost(mob/dead/observer/user)
+	if(user.client)
+		if(user.gas_scan && atmos_scan(user=user, target=src, silent=TRUE))
+			return TRUE
```
Comment (SteelSlayer) [inline code/_onclick/observer_onclick.dm:68]: Styling
```suggestion
	if(user.client && user.gas_scan && atmos_scan(user = user, target = src, silent = TRUE))
		return TRUE
```

File: code/game/objects/items/devices/scanners/gas_analyzer.dm [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+	var/cooldown_time = 250
+	var/accuracy // 0 is the best accuracy.
+	var/list/last_gasmix_data
+	var/list/history_gasmix_data
+	var/history_gasmix_index = 0
+	var/history_view_mode = ANALYZER_HISTORY_MODE_KPA
+	var/scan_range = 1
+	var/auto_updating = TRUE
+	var/target_mode = ANALYZER_MODE_SURROUNDINGS
+	var/atom/scan_target
```
Comment (SteelSlayer) [inline code/game/objects/items/devices/scanners/gas_analyzer.dm:31]: This needs to be nulled in `Destroy`

File: code/game/objects/items/devices/scanners/gas_analyzer.dm [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+			to_chat(user, "<span class='warning'>[src]'s barometer function says that the next storm will breeze on by.</span>")
+	else
+		var/next_hit = SSweather.next_hit_by_zlevel["[T.z]"]
+		var/fixed = next_hit ? next_hit - world.time : -1
+		if(fixed < 0)
+			to_chat(user, "<span class='warning'>[src]'s barometer function was unable to trace any weather patterns.</span>")
+		else
+			to_chat(user, "<span class='warning'>[src]'s barometer function says a storm will land in approximately [butchertime(fixed)].</span>")
+	cooldown = TRUE
+	addtimer(CALLBACK(src,/obj/item/analyzer/proc/ping), cooldown_time)
```
Comment (SteelSlayer) [inline code/game/objects/items/devices/scanners/gas_analyzer.dm:80]:
```suggestion
	addtimer(CALLBACK(src, PROC_REF(ping)), cooldown_time)
```

File: code/game/objects/items/devices/scanners/gas_analyzer.dm [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+		if(length(history_gasmix_data) >= ANALYZER_HISTORY_SIZE)
+			history_gasmix_data.Cut(ANALYZER_HISTORY_SIZE, length(history_gasmix_data) + 1)
+		history_gasmix_data.Insert(1, list(new_gasmix_data))
+
+/obj/item/analyzer/attack_self(mob/user)
+	if(user.stat != CONSCIOUS)
+		return
+	target_mode = ANALYZER_MODE_SURROUNDINGS
+	atmos_scan(user=user, target=get_turf(src), silent=FALSE, print=FALSE)
+	on_analyze(source=user, target=get_turf(src), save_data=!auto_updating)
```
Comment (SteelSlayer) [inline code/game/objects/items/devices/scanners/gas_analyzer.dm:186]: Spacing around the `=` please.

File: code/game/objects/items/devices/scanners/gas_analyzer.dm [hunk 9]
```diff
@@ -0,0 +1,269 @@
...
+
+/obj/item/analyzer/afterattack(atom/target, mob/user, proximity, params)
+	. = ..()
+	if(!can_see(user, target, scan_range))
+		return
+	target_mode = ANALYZER_MODE_TARGET
+	if(target == user || target == user.loc)
+		target_mode = ANALYZER_MODE_SURROUNDINGS
+	atmos_scan(user=user, target=(target.return_analyzable_air() ? target : get_turf(src)), print=FALSE)
+	on_analyze(source=user, target=(target.return_analyzable_air() ? target : get_turf(src)), save_data=!auto_updating)
```
Comment (SteelSlayer) [inline code/game/objects/items/devices/scanners/gas_analyzer.dm:198]: More spacing

**Review by SteelSlayer — CHANGES_REQUESTED** [review]

**Comment by larentoun** [comment]
No reason to support upstream if they can't do a little thing to support downstream.

**Pull request closed without merging**